core/src/muxing/boxed.rs

```diff
@@ -38,6 +39,7 @@ impl<T> StreamMuxer for Wrap<T>
 where
     T: StreamMuxer,
     T::Substream: AsyncRead + AsyncWrite + Send + Unpin + 'static,
+    T::Error: Into<Box<dyn Error + Send + Sync>>,
```

Comment:
```suggestion
    T::Error: Error + Send + Sync>,
```
Nit: Wouldn't this simpler trait bound also work, given that `Into<Box<dyn ..>>` is auto-implemented for `E: Error + Send + Sync`?

Reply:
Probably yes. I've copied the bounds from `std::io::Error::new` because that is where I am passing the type to. Omitting the `Into` here is probably okay :)

Reply:
I decided to put the `Error` bound directly on the associated type so I don't have to repeat it everywhere. The `Send + Sync` bounds stay on higher layers for consistency with other bounds.

Reply:
> I decided to put the `Error` bound directly on the associated type so I don't have to repeat it everywhere.

It is only in 4 locations, right? https://github.com/libp2p/rust-libp2p/pull/2710/commits/a479eb88ed0a8914bedbad856ff18b181b2246df I would prefer bounds to be as close to where they are used as possible. If I am not mistaken `StreamMuxer` doesn't need `type Error` to implement `Error`, right?

Reply:
It doesn't. It also doesn't necessarily need `Substream` to implement `AsyncRead + AsyncWrite` but it is a bit pointless if it doesn't.

If we remove the `Error` bounds, we should also remove `AsyncRead + AsyncWrite` for consistency but that will get a bit redundant I think.

The current style seems to be that _functional_ bounds are put on associated types and auto-traits are left off.